src/analyses/variable-sensitivity/abstract_enviroment.cpp

```diff
@@ -0,0 +1,585 @@
+
+abstract_object_pointert abstract_environmentt::abstract_object_factory(
+  const typet type, bool top, bool bottom) const
+{
+  // TODO (tkiley): Here we should look at some config file
+  if(type.id()==ID_signedbv || type.id()==ID_bool || type.id()==ID_c_bool)
+  {
+    return abstract_object_pointert(
+      new constant_abstract_valuet(type, top, bottom));
+  }
+  else
+  {
+    return abstract_object_pointert(new abstract_objectt(type, top, false));
+  }
+}
+
+/*******************************************************************\
+
+Function: abstract_environmentt::abstract_object_factory
+
+  Inputs:
+   type - the type of the object whose state should be tracked
+   expr - the starting value of the symbol
+
+ Outputs: The abstract object that has been created
+
+ Purpose: Look at the configuration for the sensitivity and create an
+          appropriate abstract_object, assigning an appropriate value
+
+\*******************************************************************/
+
+abstract_object_pointert abstract_environmentt::abstract_object_factory(
+  const typet type, const constant_exprt e) const
+{
+  assert(type==e.type());
+  if(type.id()==ID_signedbv || type.id()==ID_bool || type.id()==ID_c_bool)
+  {
+    return abstract_object_pointert(
+      new constant_abstract_valuet(e));
+  }
+  else
+  {
+    return abstract_object_pointert(new abstract_objectt(e));
+  }
+}
+
+/*******************************************************************\
+
+Function: abstract_environmentt::merge
+
+  Inputs:
+   env - the other environment
+
+ Outputs: ?
+
+ Purpose: ?
+
+\*******************************************************************/
+
+bool abstract_environmentt::merge(const abstract_environmentt &env)
+{
+  // Use the sharing_map's "iterative over all differences" functionality
+  // This should give a significant performance boost
+  // We can strip down to just the things that are in both
+
+  // for each entry in the incoming environment we need to either add it
+  // if it is new, or merge with the existing key if it is not present
+
+
+
+  if(is_bottom)
+  {
+    *this=env;
+    return !env.is_bottom;
+  }
+  else if(env.is_bottom)
+  {
+    return false;
+  }
+  else
+  {
+    bool modified=false;
+    for(const auto &entry:env.map)
+    {
+      if(map.find(entry.first)!=map.end())
+      {
+        bool object_modified=false;
+        abstract_object_pointert new_object=map[entry.first]->merge(
+          entry.second, object_modified);
+
+      modified|=object_modified;
+        map[entry.first]=new_object;
+
+      }
+
+      if(map[entry.first]->is_top())
+      {
+        map.erase(entry.first);
+        modified=true;
+#ifdef DEBUG
+      std::cout << "Removing " << entry.first.get_identifier() << std::endl;
+#endif
+      }
+    }
+
+    std::vector<map_keyt> to_remove;
+    for(const auto &entry : map)
+    {
+      if(env.map.find(entry.first)==env.map.end())
+      {
+        to_remove.push_back(entry.first);
```

Comment:
Yes of course, this is nonsense - serves me right for copying code (badly) without really considering it. Also reveals there is evidently at least one two few tests since this didn't fail.